cmd/sign-off-checker/main.go

```diff
@@ -17,129 +17,198 @@ limitations under the License.
 package main
 
 import (
-	"context"
 	"fmt"
 	"log"
+	"net"
 	"net/http"
+	"net/url"
 	"os"
-	"regexp"
+	"time"
 
 	"github.com/google/go-github/github"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
 	"golang.org/x/oauth2"
+
+	"github.com/heptiolabs/sign-off-checker/pkg/register"
+	"github.com/heptiolabs/sign-off-checker/pkg/webhook"
 )
 
-var client *github.Client
-var secret []byte
+// CLI entrypoint
+func main() {
+	rootCmd := &cobra.Command{
+		Use:     "SHARED_SECRET='[...]' GITHUB_TOKEN='[...]' sign-off-checker",
+		Short:   "A GitHub integration to ensure commits have \"Signed-off-by\".",
+		Args:    cobra.NoArgs,
+		PreRunE: func(_ *cobra.Command, _ []string) error { return validate() },
+		RunE:    func(_ *cobra.Command, _ []string) error { return run() },
+	}
 
-var testRE *regexp.Regexp
+	// $SHARED_SECRET
+	viper.BindEnv("sharedSecret", "SHARED_SECRET")
 
-func init() {
-	testRE = regexp.MustCompile(`(?mi)^signed-off-by:`)
-}
+	// $GITHUB_TOKEN
+	viper.BindEnv("githubToken", "GITHUB_TOKEN")
 
-func loggingMiddleware(handler http.Handler) http.Handler {
-	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		log.Printf("%s %s %s", r.RemoteAddr, r.Method, r.URL)
-		handler.ServeHTTP(w, r)
-	})
-}
+	// --listen / $LISTEN
+	rootCmd.Flags().String(
```

Comment:
IIRC this returns the flag, so you might be able to pass that to the BindPFlag call on the next line.

Reply:
It returns a `*string` that holds the flag's value and I can't see an easy way to define the flag and bind it into viper in a single shot. I am not a fan of the stringly typed interfaces here.

I thought about defining some constants for the viper config keys like `const sharedSecretKey = "sharedSecret"` so at least I've got a shot at catching the bug if I typo it to `"shareSecret"` somewhere.

Reply:
I took a shot at this in 481c419.